File: javascript/path_traversal/fileUpload.js
const express = require('express');
const fs = require('fs');
const path = require('path');
const AdmZip = require('adm-zip');
const multer = require('multer');

const router = express.Router();
const UPLOAD_DIR = '/var/www/uploads';
const upload = multer({ dest: UPLOAD_DIR });

// VULN 1: Write uploaded file to user-specified destination path - arbitrary file write
router.post('/upload', upload.single('file'), (req, res) => {
    const dest = req.body.destination;
    const savePath = path.join(UPLOAD_DIR, dest);
    fs.renameSync(req.file.path, savePath);
    res.send('uploaded');
});

// VULN 2: Zip extraction to user-controlled directory - zip slip vulnerability
router.post('/extract', upload.single('archive'), (req, res) => {
    const extractTo = req.body.extract_to;
    const zip = new AdmZip(req.file.path);
    zip.extractAllTo(extractTo, true);
    res.send('extracted');
});

// VULN 3: Read uploaded file using user-controlled filename - path traversal on read
router.get('/preview', (req, res) => {
    const username = req.query.user;
    const filename = req.query.file;
    const filePath = UPLOAD_DIR + '/' + username + '/' + filename;
    res.send(fs.readFileSync(filePath, 'utf8'));
});

module.exports = router;

File: go/path_traversal/file_upload.go
package pathtraversal

import (
	"archive/zip"
	"io"
	"net/http"
	"os"
	"path/filepath"
)

const uploadDir = "/var/www/uploads"

// VULN 1: Write uploaded file to user-specified destination - arbitrary file write
func UploadFile(w http.ResponseWriter, r *http.Request) {
	r.ParseMultipartForm(32 << 20)
	dest := r.FormValue("destination")
	file, _, err := r.FormFile("file")
	if err != nil {
		http.Error(w, err.Error(), 500)
		return
	}
	defer file.Close()
	savePath := filepath.Join(uploadDir, dest)
	out, _ := os.Create(savePath)
	defer out.Close()
	io.Copy(out, file)
	w.Write([]byte("uploaded"))
}

// VULN 2: Zip extraction to user-controlled directory - zip slip vulnerability
func ExtractArchive(w http.ResponseWriter, r *http.Request) {
	extractTo := r.FormValue("extract_to")
	archivePath := filepath.Join(uploadDir, r.FormValue("archive"))
	zr, _ := zip.OpenReader(archivePath)
	defer zr.Close()
	for _, f := range zr.File {
		// zip slip: f.Name may contain ../ sequences
		destPath := filepath.Join(extractTo, f.Name)
		out, _ := os.Create(destPath)
		rc, _ := f.Open()
		io.Copy(out, rc)
		rc.Close()
		out.Close()
	}
	w.Write([]byte("extracted"))
}

// VULN 3: Read file using user-controlled filename - path traversal on read
func PreviewUpload(w http.ResponseWriter, r *http.Request) {
	username := r.URL.Query().Get("user")
	filename := r.URL.Query().Get("file")
	path := uploadDir + "/" + username + "/" + filename
	data, err := os.ReadFile(path)
	if err != nil {
		http.Error(w, err.Error(), 500)
		return
	}
	w.Write(data)
}
